### DISCUSSION TITLE
Can't figure out how to clean up element event listeners

### DISCUSSION THREAD

**Discussion by NabeelAhmed1721** [solidjs/solid#1760] [Q&A] [answered]
My code looks something like this:

```javascript
/* app.tsx */
function App() {
    const [video, setVideo] = createSignal<HTMLVideoElement>();
    
    onMount(() => {
        createVideo("/test.mp4").then(setVideo).catch(console.error)
    })

    return (
        <Preview video={video()} onVideoClose={() => setVideo(null)} />
    )
}

/* preview.tsx */

type PreviewProps = { video: HTMLVideoElement; onVideoClose: () => void; }

function Preview(props: PreviewProps) {
    createEffect(() => {
               /* This works as expected. Event listeners work correctly. */
		props.video.addEventListener('timeupdate', handleTimeUpdate);
		props.video.addEventListener('play', handlePlay);
		props.video.addEventListener('pause', handlePause);
		props.video.addEventListener('seeking', handleSeeking);
		props.video.addEventListener('seeked', handleSeeked);
    });

    onCleanup(() => {
               /*
                  This doesn't work. Can't remove event listeners for clean up.
                  By the time onCleanUp runs in this component, `props.video` is `null` from `app.tsx`.
               */
        	props.video.removeEventListener('timeupdate', handleTimeUpdate);
		props.video.removeEventListener('play', handlePlay);
		props.video.removeEventListener('pause', handlePause);
		props.video.removeEventListener('seeking', handleSeeking);
		props.video.removeEventListener('seeked', handleSeeked);
    })

    return (
        <button onClick={props.onVideoClose}>remove video</button>
    )
}
```

The problem I'm having is I can't properly clean up event listeners for `props.video` because when the button to "remove video" is pressed, the `video` signal is changed to `null`, and by the time clean up happens in `preview.tsx`, `props.video` is null.

What should I do? Thank you.

**Comment by AlexErrant** [comment]
[Tutorial states](https://www.solidjs.com/tutorial/lifecycles_oncleanup?solved):

> You can call it [onCleanup] at any scope and it will run when that scope is triggered to re-evaluate and when it is finally disposed.

Since you create the event listener in the effect, you should dispose of the event listener _in that effect_.

```ts
    createEffect(() => {
		props.video.addEventListener('timeupdate', handleTimeUpdate);
		props.video.addEventListener('play', handlePlay);
		props.video.addEventListener('pause', handlePause);
		props.video.addEventListener('seeking', handleSeeking);
		props.video.addEventListener('seeked', handleSeeked);
				onCleanup(() => {
			        	props.video.removeEventListener('timeupdate', handleTimeUpdate);
						props.video.removeEventListener('play', handlePlay);
						props.video.removeEventListener('pause', handlePause);
						props.video.removeEventListener('seeking', handleSeeking);
						props.video.removeEventListener('seeked', handleSeeked);
				    })
    });
```

(I'm guessing here - haven't tested anything.)

**Reply by NabeelAhmed1721** [reply]
Hey, thank you for the response.

I've just tested it, and unfortunately, this doesn't work either.

**Reply by AlexErrant** [reply]
Hm, maybe try in the onMount? Ref: https://github.com/solidjs/solid/issues/1616#issuecomment-1465227499

**Reply by NabeelAhmed1721** [reply, marked as answer]
I was running into the same issue even when using `onMount`. However, I was able to find a solution through some experimentation.

```javascript
    createEffect(() => {
                const video = props.video; // use variable to store props.videos

		video.addEventListener('timeupdate', handleTimeUpdate);
		video.addEventListener('play', handlePlay);
		video.addEventListener('pause', handlePause);
		video.addEventListener('seeking', handleSeeking);
		video.addEventListener('seeked', handleSeeked);
		
                onCleanup(() => {
	                video.removeEventListener('timeupdate', handleTimeUpdate);
			video.removeEventListener('play', handlePlay);
			video.removeEventListener('pause', handlePause);
			video.removeEventListener('seeking', handleSeeking);
			video.removeEventListener('seeked', handleSeeked);
		 })
    });
```

By referencing `props.video` through a `video` const, I was able to get this to work. I'm not sure if this is a proper solution, but it does work. Nonetheless, the solidjs linter isn't warning me about anything so I guess I'm in the clear.

**Reply by snnsnn** [reply]
You should not use createEffect for attaching event listeners because you will have more and more listeners with each state update. You can directly add them inside a component's body or use `onMount`. `onMount` adds an effect that runs once. Your issue could not be due to onMount because there is no difference between effects that are created via onMount and createEffect.

**Reply by snnsnn** [reply]
> Since you create the event listener in the effect, you should dispose of the event listener in that effect.

Not necessarily. `onCleanup` is just a callback, as long as it has access to right variables, you can called them outside an effect, directly inside a component's body.

**Comment by snnsnn** [comment]
You don't need to create a video and pass it to preview. All you do is set path, if that is all you need, you can use the signal to store the path and pass it around. Now you can create a video using the path in your Preview component and add event listeners to the video directly without using any effects. Since video will be destroy when component is disposed, no need to remove the event listeners either.
